### PR TITLE
New lobby option to disable Manual unit share

### PR DESCRIPTION
**Pull request by Crotalus**
Make it possible to disable manual unit sharing, units are still given at death if full share is on

Options:
yes - manual share allowed (default)
no_engineers - no sharing of engineers/factories (filter out units with CONSTRUCTION or ENGINEER)
none - no manual share of units

### CODE REVIEW

File: lua/ui/lobby/lobbyOptions.lua [hunk 1]
```diff
@@ -465,6 +465,29 @@ globalOpts = {
...
+        help = "<LOC lobui_0791>Are players allowed to manually give units?",
+        key = 'ManualUnitShare',
+        values = {
+            {
+                text = "<LOC _Yes>Yes",
+                help = "Manual unit sharing are allowed",
+                key = 'all',
+            },
+            {
+                text = "Yes except builders",
```
Comment (Garanas) [inline lua/ui/lobby/lobbyOptions.lua:480]: Can you add this one too to the translation?

File: loc/US/strings_db.lua [hunk 2]
```diff
@@ -7666,6 +7666,13 @@ lobui_0781="Teams will be balanced with up to 5%% tolerance of best setup to mak
 lobui_0782="Optimal balance (Mirrored)"
 lobui_0783="Teams will be optimally balanced, mirrored start locations"
 
+lobui_0790="Manual Unit Sharing"
+lobui_0791="Are players allowed to manually give units?"
+lobui_0792="Manual unit sharing are allowed"
```
Comment (Garanas) [inline loc/US/strings_db.lua:7671]: should this be `is` instead of `are`?

File: loc/US/strings_db.lua [hunk 3]
```diff
@@ -7666,6 +7666,13 @@ lobui_0781="Teams will be balanced with up to 5%% tolerance of best setup to mak
...
+lobui_0793="Yes except builders"
```
Comment (Garanas) [inline loc/US/strings_db.lua:7672]: And I think this should be `Yes, except builders`

File: loc/US/strings_db.lua [hunk 4]
```diff
@@ -7666,6 +7666,13 @@ lobui_0781="Teams will be balanced with up to 5%% tolerance of best setup to mak
...
+lobui_0794="No manual sharing of builder/factory"
```
Comment (Garanas) [inline loc/US/strings_db.lua:7673]: and then specify them here as engineers and factories, anything that has a build rate

File: lua/SimUtils.lua [hunk 5]
```diff
@@ -556,9 +556,16 @@ function TransferUnfinishedUnitsAfterDeath(units, armies)
...
+    if manualShare == 'none' then return end
+
     if units then
         local owner = units[1].Army
         if OkayToMessWithArmy(owner) and IsAlly(owner,data.To) then
+            if manualShare == 'no_builders' then
+                units = EntityCategoryFilterDown(categories.ALLUNITS - categories.CONSTRUCTION - categories.ENGINEER, units)
+            end
+            
             TransferUnitsOwnership(units, data.To)
```
Comment (Garanas) [inline lua/SimUtils.lua:569]: There needs to be some UI element telling you if you're trying to give engineers / factories but you can't

**Comment by 4z0t** [comment]
Also, what would happen to Exps like fatboy, megabot, atlantis, czar, tempest and novax?

File: lua/SimUtils.lua (hunk 5 above)
Comment (Garanas) [reply lua/SimUtils.lua:569]: You can compare the length of units before / after if the filtering to confirm it has engineers in it

**Comment by Garanas** [comment]
I've asked the balance team. For experimentals, we can add on a per-unit bases restrictions to prevent them from being able to make engineers.

**Comment by Marc-Spector** [comment]
T_R_U_Putin is sad somewhere 😅

**Comment by Emperor-Penguin1** [comment]
If you do add restrictions for building units with T4 (which makes sense), you should probably also make it so those restrictions get removed when the units are transferred via full share.

File: loc/US/strings_db.lua (hunk 2 above)
Comment (Crotalus) [reply loc/US/strings_db.lua:7671]: Fixed

File: loc/US/strings_db.lua (hunk 3 above)
Comment (Crotalus) [reply loc/US/strings_db.lua:7672]: Fixed

File: loc/US/strings_db.lua (hunk 4 above)
Comment (Crotalus) [reply loc/US/strings_db.lua:7673]: Changed to engineers/factories (until we know exactly what will be banned)

File: lua/SimUtils.lua (hunk 5 above)
Comment (Crotalus) [reply lua/SimUtils.lua:576]: Added a print message now when not all units could be given

**Comment by Garanas** [comment]
If I am not mistaken then the consensus of the balance team is that this is not a balance related change. @Tagada14 can you confirm this?

**Comment by Garanas** [comment]
I'm merging this - I think it can use some fine tuning before the release based on feedback. In particular, you can't share some experimental units at the moment due to their building capacities.

**Pull request merged**